service/controller/v1/resource/drainer/create.go

```diff
@@ -0,0 +1,159 @@
+package drainer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apismetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/giantswarm/node-operator/service/controller/v1/key"
+)
+
+const (
+	// UnschedulablePatch is the JSON patch structure being applied to nodes using
+	// a strategic merge patch in order to drain them.
+	UnschedulablePatch = `{"spec":{"unschedulable":true}}`
+)
+
+// EnsureCreated represents the node resource implementation to manage on demand
+// node draining for guest clusters.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	drainerConfig, err := key.ToDrainerConfig(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if drainerConfig.Status.HasFinalCondition() {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "node config status already has final state")
+		resourcecanceledcontext.SetCanceled(ctx)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+		return nil
+	}
+
+	var k8sClient kubernetes.Interface
+	{
+		i := key.ClusterIDFromDrainerConfig(drainerConfig)
+		e := key.ClusterEndpointFromDrainerConfig(drainerConfig)
+		k8sClient, err = r.guestCluster.NewK8sClient(ctx, i, e)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "cordoning guest cluster node")
+
+		n := key.NodeNameFromDrainerConfig(drainerConfig)
+		t := types.StrategicMergePatchType
+		p := []byte(UnschedulablePatch)
+
+		_, err := k8sClient.CoreV1().Nodes().Patch(n, t, p)
+		if apierrors.IsNotFound(err) {
+			// It might happen the node we want to drain got already removed. This
+			// might even be due to human intervention. In case we cannot find the
+			// node we assume the draining was successful and set the node config
+			// status accordingly.
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "guest cluster node not found")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "setting node config status of guest cluster node to final state")
+
+			drainerConfig.Status.Conditions = append(drainerConfig.Status.Conditions, drainerConfig.Status.NewFinalCondition())
+
+			_, err := r.g8sClient.CoreV1alpha1().DrainerConfigs(drainerConfig.GetNamespace()).UpdateStatus(&drainerConfig)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "set node config status of guest cluster node to final state")
+			resourcecanceledcontext.SetCanceled(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
```

Comment:
General thought: the ` for custom object` seems to be redundant. We have the `object` key in the log message.

Reply:
Again, this is just copy/paste. Also, it is like this everywhere.